x/finality/types/msg.go

```diff
@@ -229,27 +229,50 @@ func (m *MsgResumeFinalityProposal) ValidateBasic() error {
 	return nil
 }
 
-func (m *MsgEquivocationEvidence) ValidateBasic() error {
+// ToEvidence converts MsgEquivocationEvidence to Evidence
+func (m *MsgEquivocationEvidence) ParseToEvidence() (*Evidence, error) {
 	if _, err := sdk.AccAddressFromBech32(m.Signer); err != nil {
-		return errorsmod.Wrapf(sdkerrors.ErrInvalidAddress, "invalid signer address (%s)", err)
-	}
-	if m.FpBtcPk == nil {
-		return ErrInvalidEquivocationEvidence.Wrap("empty FpBtcPk")
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidAddress, "invalid signer address (%s)", err)
 	}
-	if m.PubRand == nil {
-		return ErrInvalidEquivocationEvidence.Wrap("empty PubRand")
+
+	fpBtcPk, err := bbntypes.NewBIP340PubKeyFromHex(m.FpBtcPkHex)
+	if err != nil {
+		return nil, ErrInvalidEquivocationEvidence.Wrapf("invalid FP BTC PK: %s", err)
 	}
-	if len(m.CanonicalAppHash) != 32 {
-		return ErrInvalidEquivocationEvidence.Wrap("malformed CanonicalAppHash")
+
+	pubRand, err := bbntypes.NewSchnorrPubRandFromHex(m.PubRandHex)
+	if err != nil {
+		return nil, ErrInvalidEquivocationEvidence.Wrapf("invalid PubRand: %s", err)
 	}
-	if len(m.ForkAppHash) != 32 {
-		return ErrInvalidEquivocationEvidence.Wrap("malformed ForkAppHash")
+
+	canonicalAppHash, err := hex.DecodeString(m.CanonicalAppHashHex)
+	if err != nil || len(canonicalAppHash) != 32 {
+		return nil, ErrInvalidEquivocationEvidence.Wrap("malformed CanonicalAppHash")
 	}
-	if m.ForkFinalitySig == nil {
-		return ErrInvalidEquivocationEvidence.Wrap("empty ForkFinalitySig")
+
+	forkAppHash, err := hex.DecodeString(m.ForkAppHashHex)
+	if err != nil || len(forkAppHash) != 32 {
+		return nil, ErrInvalidEquivocationEvidence.Wrap("malformed ForkAppHash")
 	}
-	if m.CanonicalFinalitySig == nil {
-		return ErrInvalidEquivocationEvidence.Wrap("empty CanonicalFinalitySig")
+
+	canonicalFinalitySig, err := bbntypes.NewSchnorrEOTSSigFromHex(m.CanonicalFinalitySigHex)
+	if err != nil {
+		return nil, ErrInvalidEquivocationEvidence.Wrapf("invalid CanonicalFinalitySig: %s", err)
 	}
-	return nil
+
+	forkFinalitySig, err := bbntypes.NewSchnorrEOTSSigFromHex(m.ForkFinalitySigHex)
+	if err != nil {
+		return nil, ErrInvalidEquivocationEvidence.Wrapf("invalid ForkFinalitySig: %s", err)
+	}
+
+	return &Evidence{
+		FpBtcPk:              fpBtcPk,
+		BlockHeight:          m.BlockHeight,
+		PubRand:              pubRand,
+		CanonicalAppHash:     canonicalAppHash,
+		ForkAppHash:          forkAppHash,
+		CanonicalFinalitySig: canonicalFinalitySig,
+		ForkFinalitySig:      forkFinalitySig,
+		SigningContext:       m.SigningContext,
```

Comment:
we could also validate that `m.SigningContext` has 64 bytes (it is hex encoded 32byte hash)

Reply:
Hmm we could, but this might not improve security, given that the signature extraction takes into account the signed message and thus the signing context anyway. Also we haven't implemented the signing context in rollup BSN contract. We can revisit this afterwards. Wdyt?